source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -174,6 +174,10 @@ void Filter::onComplete(Filters::Common::ExtAuthz::ResponsePtr&& response) {
       ENVOY_STREAM_LOG(trace, "'{}':'{}'", *callbacks_, header.first.get(), header.second);
       request_headers_->setCopy(header.first, header.second);
     }
+    for (const auto& header : response->headers_to_add_and_append) {
+      ENVOY_STREAM_LOG(trace, "'{}':'{}'", *callbacks_, header.first.get(), header.second);
+      request_headers_->addCopy(header.first, header.second);
```

Comment:
I think we can replace the `appendCopy` below with `addCopy` instead.

Reply:
but this may break backwards compatibility for the original behavior of headers_to_append

Reply:
How do you know? I think this is an opportunity to add tests for it?

Reply:
See: https://github.com/envoyproxy/envoy/pull/11617 I added a test for this case. From the [RFC](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2) standpoint, I think it is OK to have multiple headers with the same key. The way to represent it as a single header is true via merging it by adding "," in-between values. (Except for, in practice, Set-Cookie, WWW-Authenticate, and Proxy-Authenticate). Hence it is only a representation problem. I lean towards `addCopy`-only since it is probably cheaper.

Reply:
but k8s may not accept comma-seperate values, it seems k8s only accept an array values as described in https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy. And in [this line](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader.go#L197), k8s writes an `allHeaderValues` function to get values from header and it does not support a comma-seperate header values as an array. So maybe add `headers_to_add_and_append` is necessary. 
If not using `headers_to_add_and_append`, we can use `headers_to_append` but remove the [`header_to_modify` check](https://gist.github.com/dio/6ea39ae3719c9701a245ed5de8dc2b8d#file-a-diff-L98). These may make sense but I'm not sure whether this may break backwards compatibility.

Reply:
OK. A couple of things. Related to modifying original request headers given a set of headers from the authorization server:

1. The current behavior requires the header in question to exist in the original request. For this particular case (solo-io/gloo#2983) we want to add new set headers that are not present in the original headers.

2. Whether we use `appendCopy` or `addCopy` for sending multiple header values with same key. The current one uses appendCopy (the implication for this is every single header, either it is an inline header or not will be concatenated as a single entry).

I agree we can keep this. But I think we can rename `headers_to_add_and_append` to `headers_to_add_copy`, while the append one can be `headers_to_append_copy` so it is easier to follow for future debugging.

Note, while it is not in the scope of this PR, that the gRPC implementation should have the corresponding new set of header value options for accommodating appending non-existent headers.

WDYT? And are you OK for adding the tests?

Reply:
It OK for me to try to add the tests. I agree with the point about renaming `headers_to_add_and_append`, but the name `headers_to_add_copy` may be confused. My point is to change `headers_to_add` to `headers_to_set` since it uses the function `set_copy`. So we can rename `headers_to_add_and_append` to `headers_to_add` for it uses `add_copy`. How do you think about the name?

Reply:
OK. Looks good. Waiting for the tests. Thanks!